### ISSUE TITLE
Some progress for RTree with --gc:arc

### ISSUE THREAD

**Issue by StefanSalewski** [nim-lang/Nim#13006]
Just tried with latest compiler:

```
$ nim -v
Nim Compiler Version 1.1.1 [Linux: amd64]
Compiled at 2020-01-01
Copyright (c) 2006-2019 by Andreas Rumpf

git hash: c3344862b0d6061cc1581f29c81b29b75c78615a
active boot switches: -d:release
```

```
$ nim c --gc:arc rtree.nim
```

```
$  ./rtree 
full nearest test
nearest: (b: [(a: -13.0, b: -7.0), (a: 0.0, b: 6.0)], l: (x: -10.0, y: 3.0, r: 3.0))
[0, 0]
TGS Bulk load 550 elements in: 8 milliseconds, 579 microseconds, and 20 nanoseconds
nearest: (b: [(a: 47, b: 59), (a: 20, b: 23)], l: 105)
Traceback (most recent call last)
/tmp/hhh/rtree.nim(1305) rtree
/tmp/hhh/rtree.nim(1277) test
/tmp/hhh/rtree.nim(146)  search
/tmp/hhh/rtree.nim(61)   s
/home/stefan/Nim/lib/system/refs_v2.nim(116) nimDestroyAndDispose
/home/stefan/Nim/lib/system/refs_v2.nim(116) nimDestroyAndDispose
...
/home/stefan/Nim/lib/system/refs_v2.nim(116) nimDestroyAndDispose
/home/stefan/Nim/lib/system/refs_v2.nim nimDecRefIsLast
Error: call depth limit reached in a debug build (2000 function calls). You can change it with -d:nimCallDepthLimit=<int> but really try to avoid deep recursions instead.
```

That looks much better than all the random crashes I got over christmas, where I tried many changes but got only crashes at other places, with strong indication for memory corruption. I also tried providing minimal examples for that crash, but minimal examples worked without problems.

RTree is hard for gc, maybe hardest part is   proc bulkLoadChunk[M, D: Dim; RT, LT](d: array[TgsX * D, seq[LTGS[D, RT, LT]]]; h: int): (H[M, D, RT, LT], Box[D, RT]) which recursively returns tuples. But all the RTree algorithm are based on detailed pseudo-code from the referenced papers, and the code works with default gc and with boehm. So it can not be that wrong. Conclusion: It will work when gc:arc is ready for production.

**Comment by Clyybber** [comment]
To give some additional context, I tried this with https://github.com/StefanSalewski/RTree and the latest #devel compiler, and the issue doesn't occur. Instead it fails with another error that doesn't look like the compilers fault.

**Comment by StefanSalewski** [comment]
Thanks for your reply. But no, same error for me with latest devel compiler:

```
$ nim -v 
Nim Compiler Version 1.1.1 [Linux: amd64]
Compiled at 2020-01-03
Copyright (c) 2006-2019 by Andreas Rumpf

git hash: fb1b51f1e37277bfa2b3c86213a6066e6b4fbcc2
active boot switches: -d:release
gcc version 9.2.0
```

```
nim c --gc:arc rtree.nim

$ ./rtree 
full nearest test
nearest: (b: [(a: -13.0, b: -7.0), (a: 0.0, b: 6.0)], l: (x: -10.0, y: 3.0, r: 3.0))
[0, 0]
TGS Bulk load 550 elements in: 10 milliseconds, 116 microseconds, and 860 nanoseconds
nearest: (b: [(a: 47, b: 59), (a: 20, b: 23)], l: 105)
Traceback (most recent call last)
/tmp/hhh/rtree.nim(1305) rtree
/tmp/hhh/rtree.nim(1277) test
/tmp/hhh/rtree.nim(146)  search
/tmp/hhh/rtree.nim(61)   s
/home/stefan/Nim/lib/system/refs_v2.nim(116) nimDestroyAndDispose
...
/home/stefan/Nim/lib/system/refs_v2.nim(116) nimDestroyAndDispose
/home/stefan/Nim/lib/system/refs_v2.nim nimDecRefIsLast
Error: call depth limit reached in a debug build (2000 function calls). You can change it with -d:nimCallDepthLimit=<int> but really try to avoid deep recursions instead.
```
All fine with default gc or boehm, but does not compile with cpp!

I guess the bulk loading is too hard for arc currently. Will try again in a few weeks.

**Comment by Clyybber** [comment]
Hmm, I can't reproduce. Do you have the latest changes in your repo pushed?

**Comment by StefanSalewski** [comment]
Strange.

```
~/RTree $ git status
On branch master
Your branch is up to date with 'origin/master'.

nothing to commit, working tree clean

```

And version 0.2 with bulk loading is there at https://github.com/StefanSalewski/RTree pushed a few months ago.

Araq told me that it may not work with 32 bit OS, but I can not test that. For my linux 64 bit all works fine with default gc and boehm.

**Comment by StefanSalewski** [comment]
But well, I guess it is a memory corruption, I had indeed very different errors when I was testing over christmas adding debugging echos or doing some small modifications.

But I did not change the code at github, for which I currently get the nimDestroyAndDispose() crash. But for you error may be different, there are many asserts in the test code, and with memory coruption various asserts may fire. Have you an idea why cpp does not compile?
